contracts/abstracted_account.algo.ts

```diff
@@ -35,35 +42,169 @@ export class AbstractedAccount extends Contract {
    */
   namedPlugins = BoxMap<bytes, PluginsKey>({ prefix: 'n' });
 
+  private assertPluginCallAllowed(app: AppID, caller: Address): void {
+    const key: PluginsKey = { application: app, allowedCaller: caller };
+
+    assert(this.plugins(key).exists, 'plugin does not exist');
+    assert(this.plugins(key).value.lastValidRound >= globals.round, 'plugin is expired');
+    assert(
+      globals.round - this.plugins(key).value.lastCalled >= this.plugins(key).value.cooldown,
+      'plugin is on cooldown'
+    );
+  }
+
+  private pluginCallAllowed(app: AppID, caller: Address): boolean {
+    const key: PluginsKey = { application: app, allowedCaller: caller };
+
+    return (
+      this.plugins(key).exists &&
+      this.plugins(key).value.lastValidRound >= globals.round &&
+      globals.round - this.plugins(key).value.lastCalled >= this.plugins(key).value.cooldown
+    );
+  }
+
+  private ensuresRekeyBack(txn: Txn): boolean {
+    if (txn.sender === this.app.address && txn.rekeyTo === this.app.address) {
+      return true;
+    }
+
+    return (
+      txn.typeEnum === TransactionType.ApplicationCall &&
+      txn.applicationID === this.app &&
+      txn.numAppArgs === 1 &&
+      txn.onCompletion === 0 &&
+      txn.applicationArgs[0] === method('arc58_verifyAuthAddr()void')
+    )
+  }
+
+  private assertRekeysBack(): void {
+    let rekeysBack = false;
+    for (let i = (this.txn.groupIndex + 1); i < this.txnGroup.length; i += 1) {
+      const txn = this.txnGroup[i];
+
+      if (this.ensuresRekeyBack(txn)) {
+        rekeysBack = true;
+      }
+    }
+
+    assert(rekeysBack, 'rekey back not found');
+  }
+
   /**
-   * Ensure that by the end of the group the abstracted account has control of itself
+   * Guarantee that our txn group is valid in a single loop over all txns in the group
+   * 
+   * @param app the plugin app id being validated
+   * @param checkGlobal whether to check the global caller for method restrictions
+   * @param checkLocal whether to check the local caller for method restrictions
    */
-  private verifyRekeyToAbstractedAccount(): void {
-    let rekeyedBack = false;
+  private assertValidGroup(app: AppID, methodOffsets: uint64[], checkGlobal: boolean, checkLocal: boolean): CallerUsed {
```

Comment:
`methodOffsets` needs to be documented. After reading the code it's also not really clear to me how it's being used

Reply:
`methodOffsets` are how we check if a method being used is in the dynamic array of methods allowed on the plugin. We require the index of the method selector in the array to be provided from the offchain call so that we can avoid having a double nested loop in the contract